### PR TITLE
GalleryAdults & Fox additions

### PR DESCRIPTION
**Pull request by SilverBeamx**
Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [ ] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [ ] Have removed `web_hi_res_512.png` when adding a new extension

This PR aims achieves 2 goals:
- GalleryAdults multisrc-lib now fetches tags from 5 pages instead of 3. This was necessary due to some sources having very few tags per page
- Implemented search for the 4 types of "currently trending" galleries which appear in the sidebar of HentaiFox's homepage

### CODE REVIEW

**Comment by SilverBeamx** [comment]
Apologies in advance for the PR title not respecting standards, i forgot to set it properly before finalizing the PR

File: src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt
Comment (vetleledaal) [inline src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt]: Maybe compare against `.encodedPath` instead.

File: src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt
Comment (vetleledaal) [inline src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt]: Can you get the CSRF token via an interceptor instead? `csrfTokenInterceptor` from WebNovel is an example. There are similar interceptors, but they are named something else.

File: src/all/hentaifox/build.gradle
Comment (vetleledaal) [inline src/all/hentaifox/build.gradle]: You are already bumping the multisrc, so no need to double bump it.

File: src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt
Comment (vetleledaal) [inline src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt]: Builder is useless here, since you just convert it back to a String.

**Review by vetleledaal — CHANGES_REQUESTED** [review]

File: src/all/hentaifox/build.gradle
Comment (SilverBeamx) [reply src/all/hentaifox/build.gradle]: Makes sense

File: src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt
Comment (SilverBeamx) [reply src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt]: WebNovel intercepts the CSRF token via an header. HentaiFox instead embeds it into the homepage itself, thus i don't think that i can intercept it the same way WebNovel does. I fetch the homepage once and then cache the token, so the traffic rate increase should be negligible

Comment (SilverBeamx) [reply src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt]: The only concern would be the Token's validity period, but in my 15 minutes of testing i have never seen it expire across multiple searches

Comment (SilverBeamx) [reply src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt]: MangaSwat uses an Interceptor with a similar usecase as this one, but all it boils down to is a wrapper around an homepage request, which is the same thing i do here. I cannot intercept it elsewhere as the CSRF token is valid only from the homepage (from what i've seen), which is normally never queried for other extension functionalities.

File: src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt
Comment (cuong-tran) [inline src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt]: Could we put this into a map then get the URI part directly instead of checking each case? Also use that map to populate `getSortOrderURIs`

File: src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt
Comment (cuong-tran) [reply src/all/hentaifox/src/eu/kanade/tachiyomi/extension/all/hentaifox/HentaiFox.kt]: csrf-token is available on other pages as well, the requestTags for example. Maybe we can get that while loading tags?

**Comment by cuong-tran** [comment]
How about this:
```
    override fun getSortOrderURIs(): List<Pair<String, String>> {
        return super.getSortOrderURIs() + sidebarCategoriesFilterStateMap.toList()
    }

    companion object {
        private val sidebarCategoriesFilterStateMap = mapOf(
            "Top Rated" to "top_rated",
            "Most Faved" to "top_faved",
            "Most Fapped" to "top_fapped",
            "Most Downloaded" to "top_downloaded",
        ).withDefault { "top_rated" }
    }
```

Then in `searchMangaRequest`:
```
sortOrderFilter?.state >= super.getSortOrderURIs().size ->
                sidebarCategoriesFilterStateMap[sortOrderFilter.state - super.getSortOrderURIs().size].second
```

Be careful when multisrc might change its `getSortOrderURIs` size in future then using specific index might fail.

**Comment by SilverBeamx** [comment]
> How about this:
> 
> ```
>     override fun getSortOrderURIs(): List<Pair<String, String>> {
>         return super.getSortOrderURIs() + sidebarCategoriesFilterStateMap.toList()
>     }
> 
>     companion object {
>         private val sidebarCategoriesFilterStateMap = mapOf(
>             "Top Rated" to "top_rated",
>             "Most Faved" to "top_faved",
>             "Most Fapped" to "top_fapped",
>             "Most Downloaded" to "top_downloaded",
>         ).withDefault { "top_rated" }
>     }
> ```
> 
> Then in `searchMangaRequest`:
> 
> ```
> sortOrderFilter?.state >= super.getSortOrderURIs().size ->
>                 sidebarCategoriesFilterStateMap[sortOrderFilter.state - super.getSortOrderURIs().size].second
> ```
> 
> Be careful when multisrc might change its `getSortOrderURIs` size in future then using specific index might fail.

Thanks for your comments. Your code gave me an idea. I used what you suggested for the Map, but instead of relying on integers, as you pointed out, i now extract the currently selected sort label from the filter, and use the special sidebar logic only if its key is present in the Map.

**Comment by SilverBeamx** [comment]
Seems ready for merge. Thanks to both of you for your insights!

**Comment by SilverBeamx** [comment]
@vetleledaal PR is waiting for you review approval. Sorry for the ping, i don't wanna pass off as insistent, i justdon't know if you've read or this PR got buried under new ones :)

**Comment by Smol-Ame** [comment]
It'll be reviewed whenever/whoever has the time, it won't get lost

**Review by vetleledaal — APPROVED** [review]

**Pull request merged**
